Refactoring the cloud orchestration check_provisioned method

pkomanek opened the pull request:
Purpose or Intent
----------------------

Refactoring the check_provisioned method for the miq_ae_orchestration based on the issue below.

Links [Optional]
-------------------
#12038

pkomanek commented:
@miq-bot add_labels automate, enhancement, refactoring, wip

gmcculloug commented:
cc @bzwei

pkomanek commented:
@miq-bot remove_label wip

bzwei commented on db/fixtures/ae_datastore/ManageIQ/Cloud/Orchestration/Provisioning/StateMachines/Methods.class/__methods__/check_provisioned.rb:
```diff
@@ -1,70 +1,113 @@
+                true
+              end
+
+              def state_and_root_vars(service)
```
This method name is not meaningful. The main logic in this method to prepare and call provider refresh.

bzwei commented on db/fixtures/ae_datastore/ManageIQ/Cloud/Orchestration/Provisioning/StateMachines/Methods.class/__methods__/check_provisioned.rb:
```diff
@@ -1,70 +1,113 @@
-  provider = service.orchestration_manager
+              def main
+                task = @handle.root["service_template_provision_task"]
+                service = task.try(:destination)
```
We want to test whether service exists and is a type of ServiceOrchestration. Log and exit if not.

bzwei commented on spec/db/fixtures/ae_datastore/ManageIQ/Cloud/Orchestration/Provisioning/StateMachines/Methods.class/__methods__/check_provisioned_spec.rb:
```diff
@@ -0,0 +1,131 @@
+    expect(ae_service.root['ae_reason']).to eq('Stack was rolled back')
+  end
+
+  it "refreshes the provider and waits for it to complete" do
```
Please group all refresh related tests in a context and share the same `before` block where you can prepare for example to return `svc_model_amazon_stack`

mkanoor commented on db/fixtures/ae_datastore/ManageIQ/Cloud/Orchestration/Provisioning/StateMachines/Methods.class/__methods__/check_provisioned.rb:
```diff
@@ -1,70 +1,119 @@
+                unless service && service.instance_of?(MiqAeMethodService::MiqAeServiceServiceOrchestration)
+                  @handle.log(:error, 'Service does not exist or has a different' \
+                                      ' type from MiqAeServiceServiceOrchestration')
+                  exit(MIQ_STOP)
```
@pkomanek 
We can't use an exit in Automate methods because when we are running in test mode, the exit will end rspec.
You can raise an exception if the service is invalid

pkomanek replied:
@mkanoor 
I know about the ending rspec tests after the getting exit, I was trying to find a way how to catch it today and this line of code `expect { described_class.new(ae_service).main }.to_not raise_error` seems working for me. I can change that from exit to raise an exception if you want. What do you think?

mkanoor replied:
@pkomanek yes lets do that.

mkanoor replied:
@pkomanek 
Should we check for the service first and then check the type of service.
Otherwise for nil we would be saying it is a different kind of service

pkomanek commented:
@mkanoor 
I split the service control check into 2 different checks for `Service is nil` and `Service has a different type from MiqAeServiceServiceOrchestration`. I also changed the `exit(MIQ_STOP)` to raise one of these exceptions and added 2 new tests to spec for checking correct raising of these exceptions.
